stroyprombeton/migrations/0019_catalog_data.py

```diff
@@ -0,0 +1,127 @@
+]
+
+
+class CreateOptionsOperation(Operation):
+    def state_forwards(self, app_label, state):
+        pass
+
+    @transaction.atomic()
+    def database_forwards(self, app_label, schema_editor, _, to_state):
+        product_model = to_state.apps.get_model(app_label, 'Product')
+        option_model = to_state.apps.get_model(app_label, 'Option')
+
+        grouped_products = groupby(
+            product_model.objects.all(),
+            key=attrgetter('name')
+        )
+        for name, products in grouped_products:
+            products = list(sorted(products, key=attrgetter('id')))
+            # this product will only left in DB
+            main_product = products[0]
+            for product in products:
+                option = option_model(
+                    product=main_product,
+                    **{f: getattr(product, f) for f in OPTION_FIELDS},
+                    **{f: getattr(product, f) for f in TAG_FIELDS}
+                )
+                option.save()
+
+    def database_backwards(self, app_label, schema_editor, from_state, _):
+        raise NotImplementedError()
+
+
+class CleanProductsOperation(Operation):
+    """Safely remove product duplicates."""
+    def state_forwards(self, app_label, state):
+        pass
+
+    @transaction.atomic()
+    def database_forwards(self, app_label, schema_editor, _, to_state):
+        product_model = to_state.apps.get_model(app_label, 'Product')
+
+        grouped_products = groupby(
+            product_model.objects.all(),
+            key=attrgetter('name')
+        )
+        for name, products in grouped_products:
+            products = list(sorted(products, key=attrgetter('id')))
+            for product in products[1:]:
+                product.delete()
+
+    def database_backwards(self, app_label, schema_editor, from_state, _):
+        raise NotImplementedError()
+
+
+class CreateTagsOperation(Operation):
+    """Move data from product columns to tag records."""
+
+    reduces_to_sql = True
+    reversible = True
+
+    def state_forwards(self, app_label, state):
+        pass
+
+    @transaction.atomic()
+    def database_forwards(self, app_label, schema_editor, _, to_state):
+        # for every map element create new TagGroup with order value
+        group_model = to_state.apps.get_model(app_label, 'TagGroup')
+        tag_model = to_state.apps.get_model(app_label, 'Tag')
+        option_model = to_state.apps.get_model(app_label, 'Option')
+
+        field_to_group = {
+            field: group_model(name=group, position=pos)
+            for pos, (field, group) in enumerate(FIELD_TO_TAG_GROUP.items())
+        }
+        for group in field_to_group.values():
+            group.save()
+
+        for option in option_model.objects.all():
+            for field, group in field_to_group.items():
+                value = getattr(option, field)
+                if value:
+                    tag_model.objects.get_or_create(name=value, group=group)
```

Comment:
We need to bind tags with options, so add created tag to the `option.tags`

Reply:
sorry, forgot to do it